core/src/dsnp/graph_page.rs

```diff
@@ -1,44 +1,158 @@
+			DsnpUserPrivateGraphChunk::try_from(payload.as_slice())?;
+		let mut decrypted_chunk: Option<Vec<u8>> = None;
+
+		// First try the key that was indicated in the page
+		if let Some(indicated_key) = keys.iter().find(|k| k.key_id == key_id) {
+			if let Ok(data) =
+				E::decrypt(&encrypted_compressed_private_graph, &indicated_key.private_key)
+			{
+				decrypted_chunk = Some(data);
+			}
+		}
+
+		// If we couldn't decrypt with (or find) the indicated key, try all keys
+		decrypted_chunk = match decrypted_chunk {
+			Some(_) => decrypted_chunk,
+			None => keys
+				.iter()
+				.find_map(|k| E::decrypt(&encrypted_compressed_private_graph, &k.private_key).ok()),
+		};
+
+		match decrypted_chunk {
+			None => Err(Error::msg("Unable to decrypt private graph chunk with any existing keys")),
+			Some(data) => {
+				let uncompressed_chunk = DeflateCompression::decompress(&data)?;
+				let connections = SchemaHandler::read_inner_graph(&uncompressed_chunk)?;
+
+				Ok(GraphPage {
+					page_id: *page_id,
+					privacy_type: PrivacyType::Private,
+					content_hash: *content_hash,
+					prids,
+					connections,
+				})
+			},
+		}
+	}
+}
+
+impl TryFrom<GraphPage> for DsnpUserPublicGraphChunk {
+	type Error = Error;
+	fn try_from(data: GraphPage) -> Result<Self, Self::Error> {
+		let uncompressed_public_graph = SchemaHandler::write_inner_graph(&data.connections)?;
+		let compressed_public_graph = DeflateCompression::compress(&uncompressed_public_graph)?;
+		Ok(Self { compressed_public_graph })
+	}
+}
+
+type PageMap = HashMap<PageId, GraphPage>;
+type GraphMap = HashMap<ConnectionType, Graph>;
 /// Graph structure to hold pages of connections of a single type
 #[derive(Debug, Clone)]
 pub struct Graph {
-	pages: HashMap<PageId, GraphPage>,
+	pub connection_type: ConnectionType,
+	pages: PageMap,
 }
 
 /// Structure to hold all of a User's Graphs, mapped by ConnectionType
 #[derive(Debug, Clone)]
 pub struct UserGraph {
-	graphs: HashMap<ConnectionType, Graph>,
+	user_id: DsnpUserId,
+	graphs: GraphMap,
+	pub update_tracker: UpdateTracker,
 }
 
 impl UserGraph {
 	/// Create a new, empty UserGraph
-	pub fn new() -> Self {
+	pub fn new(user_id: &DsnpUserId) -> Self {
 		Self {
-			graphs: HashMap::<ConnectionType, Graph>::from([
-				(ConnectionType::Follow(PrivacyType::Public), Graph::new()),
-				(ConnectionType::Follow(PrivacyType::Private), Graph::new()),
-				(ConnectionType::Friendship(PrivacyType::Public), Graph::new()),
-				(ConnectionType::Friendship(PrivacyType::Private), Graph::new()),
+			user_id: *user_id,
+			graphs: GraphMap::from([
+				(
+					ConnectionType::Follow(PrivacyType::Public),
+					Graph::new(ConnectionType::Follow(PrivacyType::Public)),
+				),
+				(
+					ConnectionType::Follow(PrivacyType::Private),
+					Graph::new(ConnectionType::Follow(PrivacyType::Private)),
+				),
+				(
+					ConnectionType::Friendship(PrivacyType::Public),
```

Comment:
I think we can remove this for now since we would also need to implement friendship verification